sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,194 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_rv_timer.h"
+
+#include <stddef.h>
+
+#include "rv_timer_regs.h"  // Generated.
+
+/// Converts an offset for a per-hart register at hart 0 to an offset
+/// at hart |hart|.
+static ptrdiff_t offset_by_hart(ptrdiff_t reg_offset, uint32_t hart) {
+  return reg_offset + 0x100 * hart;
+}
+
+dif_rv_timer_init_result_t dif_rv_timer_init(dif_rv_timer_config_t config,
+                                             dif_rv_timer_t *timer_out) {
+  if (timer_out == NULL) {
+    return kDifRvTimerInitBadArg;
+  }
+
+  uint32_t cfg_value = 0;
+  cfg_value |= (config.tick_params.prescale & RV_TIMER_CFG0_PRESCALE_MASK)
+               << RV_TIMER_CFG0_PRESCALE_OFFSET;
+  cfg_value |= (config.tick_params.tick_step & RV_TIMER_CFG0_STEP_MASK)
+               << RV_TIMER_CFG0_STEP_OFFSET;
+  mmio_region_write32(config.base_addr,
+                      offset_by_hart(RV_TIMER_CFG0_REG_OFFSET, config.hart),
+                      cfg_value);
+
+  timer_out->base_addr = config.base_addr;
+  timer_out->hart = config.hart;
+
+  return kDifRvTimerInitOk;
+}
+
+// A simple implementation of the Euclidean Algorithm by repeated remainder.
+static uint64_t euclidean_gcd(uint64_t a, uint64_t b) {
+  while (b != 0) {
+    uint64_t old_b = b;
+    b = a % b;
+    a = old_b;
+  }
+
+  return a;
+}
+
+dif_rv_timer_approximate_tick_params_result_t
+dif_rv_timer_approximate_tick_params(uint64_t clock_freq, uint64_t counter_freq,
+                                     dif_rv_timer_tick_params_t *out) {
+  if (out == NULL) {
+    return kDifRvTimerApproximateTickParamsBadArg;
+  }
+
+  // We have the following relation:
+  //   clock_freq * (step / (prescale + 1)) = counter_freq
```

Comment:
I think it could be easier to understand if written as `counter_freq = clock_freq / (prescale + 1) * step`

Reply:
I want to keep the fraction, to make it clear that step/prescale form a conversion factor together.